### PR TITLE
Add failure, derive fail for error kind

### PR DESCRIPTION
**Pull request by gadomski**
This implements `std::error::Error` for `ErrorKind` as well.

### CODE REVIEW

**Comment by Eh2406** [comment]
Is this an api breaking change?

**Comment by mrhooray** [comment]
@gadomski could you share your use cases? looks like this adds derived implementation for `Fail` and `Display` traits to `ErrorKind` so it should be backward compatible, could you confirm on this?

**Comment by gadomski** [comment]
> could you share your use cases?

I was using **kdtree-rs** in a downstream that does implement [`Error`](https://doc.rust-lang.org/std/error/index.html), and I either:

a. Had to create a wrapper for `kdtree::ErrorKind` that implemented `Error`, or
b. Implement `Error` back in **kdtree-rs** itself.

While there isn't a mandate to derive `Error` for everything, as a crate that'll be used in a lot of other crates it'll help others use your crate.

[failure](https://github.com/withoutboats/failure) is self-described as a eventual replacement for `std::error::Error` and can be a drop-in replacement for it now. I thought it's not a bad idea to jump on that train early.

> it should be backward compatible, could you confirm on this?

Uh, I agree that it should be, but haven't tested it specifically. Your tests didn't break so that's a start? :-)

**Comment by mrhooray** [comment]
@gadomski after reading through this [stack overflow question](https://stackoverflow.com/questions/41185023/what-exactly-is-considered-a-breaking-change-to-a-library-crate), [api evolution rfc](https://github.com/nox/rust-rfcs/blob/master/text/1105-api-evolution.md) and [rebalancing coherence rfc](https://github.com/rust-lang/rfcs/blob/master/text/1023-rebalancing-coherence.md), @Eh2406's point on potential breaking change is valid.

According to [trait implementation section](https://github.com/nox/rust-rfcs/blob/master/text/1105-api-evolution.md#trait-implementations): implementing any non-fundamental trait is a minor change and `technically implementing any existing trait is a breaking change!`. This kind of breakage should be less of an issue given this is a pre-1.0 crate.

My concern is that it seems a type cannot implement both `std::error::Error` and `failure::Fail`(got conflicting implementations error) thus merging this pr would prevent us implementing the std interface which is probably more widely used(though with less functionalities compared to failure). 

Would you suggest us implementing `std::error::Error` instead? Optionally we can have a `to_failure` function for conversation if possible.

**Comment by gadomski** [comment]
> This kind of breakage should be less of an issue given this is a pre-1.0 crate.

I agree with this.

> Would you suggest us implementing std::error::Error instead?

If you're trying to keep dependencies low, then yes. From the **failure** docs:

> Additionally, all types that already implement std::error::Error, and are also Send, Sync, and 'static, implement Fail by a blanket impl.

As a relatively low-level library, it might make sense to just implement `std::error::Error`. I find a helper library such as [quick-error](https://github.com/tailhook/quick-error) or, now, **failure** to earn its keep as libraries grow, but if things stay small a manual error implementation might make sense for you. I've opened #14 in case you want to go that way.

File: Cargo.toml [hunk 1]
```diff
@@ -9,5 +9,9 @@ repository = "https://github.com/mrhooray/kdtree-rs"
 documentation = "https://docs.rs/kdtree"
 license = "MIT OR Apache-2.0"
 
+[dependencies]
+failure = "~0.1"
+failure_derive = "~0.1"
```
Comment (CryZe) [inline Cargo.toml]: Tilde Requirements like these are questionable. Use just 0.1 if you want better semver compatibility.

Comment (gadomski) [reply Cargo.toml]: Sorry, was just copying the tilde use in `dev-dependencies`. Updated and force-pushed.

**Comment by mrhooray** [comment]
thanks @gadomski , merged #14, hope it makes the crate easier to use in use cases you mentioned

**Pull request closed without merging**